sklearn/preprocessing/value_dropper.py

```diff
@@ -0,0 +1,216 @@
+
+    """
+    X, y = check_X_y(X, y)
+
+    if missing_mask_only and not return_missing_mask:
+        raise ValueError("Both missing_mask_only and return_missing_mask"
+                         "cannot be True")
+
+    if missing_fraction >= 1:
+        raise ValueError("The missing_fraction cannot be greater than"
+                         " or equal to 1.")
+
+    if copy:
+        X = X.copy()
+        if missing_mask is not None:
+            missing_mask = missing_mask.copy()
+
+    if (isinstance(missing_values, str) and
+            missing_values.lower() == "NaN"):
+        missing_values = np.nan
+
+    if missing_mask is None:
+        if np.isnan(missing_values):
+            missing_mask = np.isnan(X)
+        else:
+            missing_mask = X == missing_values
+
+    n_samples, n_features = X.shape
+    n_elements = n_samples * n_features
+
+    current_n_missing = np.count_nonzero(missing_mask)
+    required_n_missing = int(missing_fraction * n_elements)
+
+    if current_n_missing > required_n_missing:
+        raise ValueError("There are currently %d missing values, "
+                         "which is >= a fraction of %0.2f that is"
+                         "expected to be missing."
+                         % (current_n_missing, missing_fraction))
+
+    rng = check_random_state(random_state)
+    n_more_missing = required_n_missing - current_n_missing
+
+    unique_labels = np.unique(y)
+    n_unique_labels = len(unique_labels)
+
+    if labels is None:
+        # Labels is an int specifying the no of labels to correlate
+        # with
+
+        if n_labels > n_unique_labels:
+            raise ValueError("The n_labels (%d) is greater than"
+                             " no of unique labels in y (%d)"
+                             % (n_labels, n_unique_labels))
+
+        labels = rng.choice(n_unique_labels, n_labels, replace=False)
+        # Reset the RNG as we don't want this operation to affect
+        # the random selection
+        rng = check_random_state(random_state)
+
+    # Filter based on labels
+    n_correlated_missing = int(n_more_missing * label_correlation)
+    n_non_correlated_missing = n_more_missing - n_correlated_missing
+    if verbose:
+        print("There are %d correlated and %d non-correlated new"
+              "missing values introduced."
+              % (n_correlated_missing, n_non_correlated_missing))
+
+    label_mask = np.zeros(missing_mask.shape, dtype=bool)
+    if label_correlation != 0:
+        for label in labels:
+            label_mask[y==label] = True
+
+    # The logic of MCAR/MNAR is implemented here
+    inv_missing_mask = ~missing_mask
+    corr_available = inv_missing_mask & label_mask
+    uncorr_available = inv_missing_mask & ~label_mask
+
+    n_corr_available = np.count_nonzero(corr_available)
+    n_uncorr_available = np.count_nonzero(uncorr_available)
+
+    n_available = n_corr_available + n_uncorr_available
+
+    if n_available < n_more_missing:
```

Comment:
Well this case could be avoided if you defined `required_n_missing = int(missing_fraction * (n_elements - current_n_missing))` instead.